proposed/2022/version-overrides.md

```diff
@@ -0,0 +1,119 @@
+
+## Summary
+
+<!-- One-paragraph description of the proposal. -->
+When specifying package versions in a central location, developers want the to be able to override versions for a particular project.  This is useful when a project is not directly tied to the rest of the repo and needs to use a newer version of a package.  Additionally, some projects have a need to specify an older version of a package when targeting an older version of .NET Framework.
+
+This proposal introduces the ability to override a centrally defined package version on a `<PackageReference />` item in a well documented and supported way.
+
+## Motivation 
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+When interviewing a number of developers who currently use the NuGet provided central package management feature, we found that the majority of them were frustrated with the lack of a way to override a centrally defined package version for a particular project. Many of these individuals would use the MSBuildSdks [Microsoft.Build.CentralPackageVersions](https://github.com/microsoft/msbuildsdks) as an alternative to give them the flexibility they needed in their repositories.
+
+The concept of dependency resolution overrides exists in many ecosystems and for NuGet, we want to ensure that there is a formal way to override package versions in the case that:
+
+- A project exists in a repository but is not part of the main set of shipping assemblies and is allowed to use newer, untested package depedencies (such as an experimental tool)
+- A project that has more advanced requirements and cannot use a single version of a package dependency for all of its target frameworks
+
+In these cases, a project can override the centrally defined package version.  However, if that project is part of a transitive graph with version conflicts, a user would still need to resolve those conflicts before restore would succeed.  The main point of this feature is to give a well documented and supported way of overriding a package version for a particular project, not to make graph resolution easier if package conflicts exist.  
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+First a user opts-in to Central Package Management with a `Directory.Packages.props` file in their directory tree.
+
+A user would specify package versions in `Directory.Packages.props`:
+
+```xml
+<Project>
+  <ItemGroup>
+    <PackageVersion Include="PackageA" Version="1.0.0" />
+    <PackageVersion Include="PackageB" Version="2.0.0" />
+  </ItemGroup>
+<Project>
+```
+
+A project would override the version with the `VersionOverride` metadata on a `<PackageReference />` item:
```

Comment:
Why is a special attribute needed for the feature? Why having just

    <PackageReference Include="PackageA" Version="3.0.0" />

would not suffice for the override?

Reply:
Unfortunately we need the ability to differentiate between accidental and deliberate actions by users.  When opting into using central package management, you do so by declaring `<PackageReference />` items with only a package ID in projects and then `<PackageVersion />` items in a single location for the whole repository.  NuGet will then tell you that you have let a package version creep into a project which means you are no longer centrally managing your package versions.

Since the `Version` metadata generates an error when using central package management, we'll need another attribute to convey that the user wants to override the centrally defined version and not emit an error since this is intentional.